### PR TITLE
Wav2Vec2 models must either throw or deal with add_apater

### PR DESCRIPTION
**Pull request by FremyCompany**
Using `Wav2Vec2Model` models with `config.add_adapter=True` as `Wav2Vec2ForCTC` (for instance) doesn't throw on creation, but eventually fails due to mismatch in the output dimension (`config.output_hidden_size`) and the `lm_head` layer put on top of it (`config.output_hidden`).

# What does this PR do?

I propose:

  1. to fix this for two sub-models that only rely on the last output (by using `output_hidden_size` when `add_adapter` is set), and 
  2. to raise an exception for the other types of models (that try to average multiple layers). Raising an explicit error is better than failing during the forward pass.

Another option is:

  3. to always throw if add_adapter is set.

## Before submitting
/!\ This is a draft pull request for review of the general proposal. I did not test this code or write tests yet. I edited the file online directly on github.

## Who can review?

Cc @patrickvonplaten since this is related to the robust-speech-event (community event)

### CODE REVIEW

**Comment by FremyCompany** [comment]
Based on the current CI results:
1. This needs reformatting if we want to merge.
2. All the Wav2Vec2-like models should mirror this change.
3. Ideally, a test would be written to ensure this doesn't regress.

But I still want to have feedback on the preferred approach before considering to go through all the hoops.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -1804,6 +1804,7 @@ class Wav2Vec2ForSequenceClassification(Wav2Vec2PreTrainedModel):
     def __init__(self, config):
         super().__init__(config)
 
+        if config.add_adapter: raise ValueError("Audio frame classification does not support the use of Wav2Vec2 adapters (config.add_adapter=True)")
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1807]:
```suggestion
        if config.add_adapter: 
            raise ValueError("Audio frame classification does not support the use of Wav2Vec2 adapters (config.add_adapter=True)")
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -1923,6 +1924,7 @@ class Wav2Vec2ForAudioFrameClassification(Wav2Vec2PreTrainedModel):
...
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1927]:
```suggestion
        if config.add_adapter: 
            raise ValueError("Audio frame classification does not support the use of Wav2Vec2 adapters (config.add_adapter=True)")
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -1687,7 +1687,7 @@ def __init__(self, config):
                 "instantiate the model as follows: `Wav2Vec2ForCTC.from_pretrained(..., vocab_size=vocab_size)`. "
                 "or define `vocab_size` of your model's configuration."
             )
-        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size)
+        self.lm_head = nn.Linear(config.output_hidden_size if config.add_adapter else config.hidden_size, config.vocab_size)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1690]:
```suggestion
        output_hidden_size = config.output_hidden_size if config.add_adapter else config.hidden_size
        self.lm_head = nn.Linear(output_hidden_size, config.vocab_size)
```

It's a bit more readable

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 4]
```diff
@@ -1599,7 +1599,7 @@ def __init__(self, config):
 
         self.wav2vec2 = Wav2Vec2Model(config)
         self.dropout = nn.Dropout(config.final_dropout)
-        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size)
+        self.lm_head = nn.Linear(config.output_hidden_size if config.add_adapter else config.hidden_size, config.vocab_size)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1602]: This class is deprecated - no need to make the change here IMO

**Comment by patrickvonplaten** [comment]
Hey @FremyCompany,

Thanks for the PR. I think it's a welcoming change. Could you run:

```make fix-copies``` so that this change is automatically copied to other varients of Wav2Vec2?

**Comment by FremyCompany** [comment]
There is an issue with the changes propagated by the fix-copies command.
```
AttributeError: 'UniSpeechSatConfig' object has no attribute 'add_adapter'
```

**Comment by FremyCompany** [comment]
I went for an `hasattr` check. 
(Assumption: if the key is absent, it defaults to False).

File: tests/test_modeling_wav2vec2.py [hunk 5]
```diff
@@ -202,6 +202,17 @@ def create_and_check_model_with_adapter(self, config, input_values, attention_ma
...
 
+    def create_and_check_model_with_adapter_for_ctc(self, config, input_values, attention_mask):
+        config.add_adapter = True
+        config.output_hidden_size = 2 * config.hidden_size
+        model = Wav2Vec2ForCTC(config=config)
+        model.to(torch_device)
+        model.eval()
+        result = model(input_values, attention_mask=attention_mask)
+        self.parent.assertEqual(
+            result.last_hidden_state.shape, (self.batch_size, self.adapter_output_seq_length, self.vocab_size)
```
Comment (patrickvonplaten) [inline tests/test_modeling_wav2vec2.py:213]:
```suggestion
            result.logits.shape, (self.batch_size, self.adapter_output_seq_length, self.vocab_size)
```

**Review by patrickvonplaten — COMMENTED** [review]
Awesome, I think one final fix and we can merge this PR

**Review by patrickvonplaten — APPROVED** [review]

**Pull request merged**